Fix soundness and optimize divrem

emmorais opened the pull request:
As @gabriel-barrett pointed out, we need to be careful while constraining Division with Remainder since the input `a` and `b` are field elements. 

When `a` and `b` are unsigned integers, we just need to check `0 <= rem < b`, and this condition is the classical condition present in all algebra books.  

However, in a field we need to check the quotient is well behaved, otherwise an adversary could choose a quotient that when multiplied by `b` gives him an arbitrary number **of his choice**, therefore allowing him to choose an invalid remainder as well.

In LEM we introduced a different way to non-deterministically check u64 range, which is more efficient. This way we avoided the soundness problem by enforcing u64 three times (`div`, `rem` and `b - rem`). 

By reusing the same ideas as in LEM we solved the soundness problem in the original hand-made circuit. We also optimized it, since three non-deterministic u64 range-checks  is still more efficient than one `to_bits_le_strict` or even `to_bits_le`.  

Function `enforce_less_than_bound` would only work for positive input (allowing the attacker to play with negative numbers to find the malicious quotient). Since it was only used for that purpose, and was unsound even for that specific usage, it was removed.

porcuquine commented:
It's good to see this trick making its way here at last, and also great that the soundness fix reduces constraints.

Looking at the implementation of `enforce_u64_div_mod`, it looks like there are still issues though. The starting point is noticing that we create new u64 versions of  the inputs (`arg1_u64` and `arg2_u64`), then *allocate* using those same values (https://github.com/lurk-lab/lurk-rs/pull/664/files#diff-7afbe1a682416c0bc156ad4957f239403a3795a4fee30b3a025528c63b094e9bR4993-R5000) but without constraining them to have a relationship to the input arguments from which they are derived: https://github.com/lurk-lab/lurk-rs/pull/664/files#diff-7afbe1a682416c0bc156ad4957f239403a3795a4fee30b3a025528c63b094e9bR4993-R5000. In fact, I *believe* it is an invariant of this function that `arg1` and `arg2` are already known to be in the u64 range. In that case, we should just be able to use the original values (`arg1.hash()` and `arg2.hash()`), right?

If the inputs have not been checked to be in range, then either:
a) It is not necessary, per the math.
b) We need to add constraints enforcing that they are u64.

Either way, I think what we need is:
- Very clear documentation of this function's invariants. What are the requirements of the input args, and clear documentation that it is being callled correctly.
- Very clear documentation of the function's logic. This might be little more than the explanation in the PR's description, which is not currently explained in the code.
- A very careful look at the whole function and how it is used in context, with an eye to ensuring that it is definitely sound – and that the code + documentation together allow an unfamiliar reader to reach that conclusion with high confidence.

It's possible that I'm missing something, in which case please let me know. I think it's also entirely possible that there are more issues beyond what I noted because I've stopped at the first point of apparently clear problems.

It's a bit worrisome to see point-fixes intended to address soundness problems — that don't even take into context the entire functional unit they intend to fix.

As a team, we need to get better at avoiding soundness bugs (very hard!), and part of that means better process — in writing, documenting, and reviewing code. We all need to develop a knee-jerk reaction to carefully examining potentially unconstrained allocations — which is one major source of problem and appears to be exhibited here.

porcuquine reviewed (changes requested):
Changes needed, as noted in my previous comment.

emmorais commented:
I added a failing unit test because we did not get a continuation error in the circuit, as in eval, when the input arguments are not u64s. 

Then I fixed it, doing the same as in eval. Therefore we are checking both arguments are u64, but it doesn't happen inside the `enforce_u64_div_mod` function.

porcuquine commented:
I am sorry, but this seems not to have addressed the primary issue in my comment. Looking back, it seems that the links to code I pasted were jumbled, so maybe that's the reason. Let me try again to point to the exact code, but if for some reason the links aren't clear — please just read the text of my original comment for a clear description of the problem.

First the args are 'coerced' to u64 — but outside the circuit: https://github.com/lurk-lab/lurk-rs/blob/92d0284994d6f4e66e556afc5d0d3b6a34b7c4dc/src/circuit/circuit_frame.rs#L5081-L5088

Then these same values are used to initialize new `AllocatedNum`s: https://github.com/lurk-lab/lurk-rs/blob/92d0284994d6f4e66e556afc5d0d3b6a34b7c4dc/src/circuit/circuit_frame.rs#L5100-L5103

These values are never constrained to have any relationship to the allocated inputs to the function from whom their value was derived.

In a certain sense, the new documentation is very accurate:

```
// On the other hand, this function never produces a circuit that
// is not satisfiable, because it internally coerces both input
// arguments to be u64s respecting equation (1).
```

What I mean is that it's true that 'this function never produces a circuit that is not satisfiable' (for this reason, at least). However, that is because the circuit is underconstrained. We are just setting the values to the 'correct' ones for later constraints, so of course it cannot fail.

Here's what I *think* I understand:
- The args are type-checked to be labeled as u64 outside this function.
- But the actual values aren't checked outside the function.
- I think what we need is to use the '64-bit construction' trick to prove that the supplied allocated values really are u64s.
- Having done that, we can (and must) then use *those* allocated values, rather than allocate new ones.

Hopefully this makes sense.

porcuquine reviewed (changes requested):
Blocking changes noted in comment.

emmorais commented:
The way I interpreted it was: "It is OK not to check the range of arg1 and arg2, because we can simply ignore the non-interesting most significant bits. It is part of the Lurk language definition that if you write `18446744073709551617u64`, then behind the scenes we are going to interpret it as `1u64`, since 18446744073709551617 = 2ˆ64 + 1". This would be a possible solution, and there would be no soundness problem because it is just part of Lurk definition. Like, if an attacker wants to use those most significant bits, no problem, will ignore it because the tag is u64, and the output of such a Lurk program would be the same as if those bits were zero. 

But I understand your concern and I see by your comment that Lurk is not defined in this way. If we get 18446744073709551617u64 in the circuit, then something went wrong with the work behind the scenes that was expected to happen, which may be the result of adversarial/malicious behavior. Therefore is not a valid witness. I added two more calls to `implies_u64` to solve this issue. Thanks for the feedback.

emmorais commented:
I also see that the main point was the unconstrained new variables. But now they were removed.

huitseeker commented:
Closing in favor of the LEM -> Lurk migration.
